### PR TITLE
Гусейнов Эмил. Задача 1. Вариант 26. Проверка лексикографической упорядоченности двух строк.

### PR DESCRIPTION
**Pull request by emilien-gus**
Последовательное решение:
Проверяем символы с одинаковыми индексами в словах на неравенство. Если окажется, что i-ый символы не совпадают, то, в случае если символ в первом слове лексиграфически больше символа во втором слове, то первое слово больше, иначе - первое слово меньше. Мы движемся по словам с помощью цыкла, пока не достигнем конца наименьшего слова. Если находим несоответсвие в символе, то записывыем результат и выходим из цыкла. Если цикл пройден полностью, то возможно 3 варианта. Первый - слова одинаковы(длины слов равны), второй - первое слово является префиксом второго слова(первое слово меньше), третий - второе слово префикс первого(первое слово больше).

Паралельное решение:
Каждый поток получает свои отрезки слов, делают проверки в циклах, потом собираем отдельные результаты в массив, где соблюдается нумерация потоков. В этом массиве сохранены локальные результаты проверки каждого сегмента. Двигаясь по этому массив ищем локальные результаты, указывающие на отличия в сегменте. Если находим такой результат, то заканчиваем поиск и сохраняем рузультат. Если же все участки совпадают, то проверяем слова на, то является ли одно слово префиксом другого. Проверка на префикс делается с помощью сравнение длин слов.

### CODE REVIEW

**Comment by nCUXe** [comment]
add TEST with empty string

File: tasks/seq/guseynov_e_check_lex_order_of_two_string/src/ops_seq.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+bool guseynov_e_check_lex_order_of_two_string_seq::TestTaskSequential::validation() {
+  internal_order_test();
+  return taskData->inputs_count[0] == 2 && taskData->outputs_count[0] == 1;
+}
+
+bool guseynov_e_check_lex_order_of_two_string_seq::TestTaskSequential::run() {
+  internal_order_test();
+  for (size_t i = 0; i < std::min(input_[0].size(), input_[1].size()); i++) {
```
Comment (Tanya-Yasakova) [inline tasks/seq/guseynov_e_check_lex_order_of_two_string/src/ops_seq.cpp:29]: Remove length into a separate variable

File: tasks/mpi/guseynov_e_check_lex_order_of_two_string/src/ops_mpi.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+bool guseynov_e_check_lex_order_of_two_string_mpi::TestMPITaskSequential::validation() {
+  internal_order_test();
+  // check count of words and count of elements of output
+  return taskData->inputs_count[0] == 2 && taskData->outputs_count[0] == 1;
+}
+
+bool guseynov_e_check_lex_order_of_two_string_mpi::TestMPITaskSequential::run() {
+  internal_order_test();
+  for (size_t i = 0; i < std::min(input_[0].size(), input_[1].size()); i++) {
```
Comment (Tanya-Yasakova) [inline tasks/mpi/guseynov_e_check_lex_order_of_two_string/src/ops_mpi.cpp:28]: Remove length into a separate variable

File: tasks/mpi/guseynov_e_check_lex_order_of_two_string/include/ops_mpi.hpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <gtest/gtest.h>
+
+#include <boost/mpi/collectives.hpp>
```
Comment (Tanya-Yasakova) [inline tasks/mpi/guseynov_e_check_lex_order_of_two_string/include/ops_mpi.hpp:5]: Remove some

File: tasks/seq/guseynov_e_check_lex_order_of_two_string/src/ops_seq.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#include "seq/guseynov_e_check_lex_order_of_two_string/include/ops_seq.hpp"
+
+#include <thread>
```
Comment (Tanya-Yasakova) [inline tasks/seq/guseynov_e_check_lex_order_of_two_string/src/ops_seq.cpp:3]: remove

File: tasks/seq/guseynov_e_check_lex_order_of_two_string/func_tests/main.cpp
[the remark this answers is not in the document]
Comment (emilien-gus) [reply tasks/seq/guseynov_e_check_lex_order_of_two_string/func_tests/main.cpp:1]: fixed

**Review by Tanya-Yasakova — APPROVED** [review]

**Review by Tanya-Yasakova — APPROVED** [review]

**Review by Tanya-Yasakova — APPROVED** [review]

**Review by Tanya-Yasakova — APPROVED** [review]

**Review by Tanya-Yasakova — APPROVED** [review]

**Review by nCUXe — APPROVED** [review]

File: tasks/mpi/guseynov_e_check_lex_order_of_two_string/src/ops_mpi.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  local_input_2_ = std::vector<char>(delta);
+  if (world.rank() == 0) {
+    local_input_1_ = std::vector<char>(input_[0].begin(), input_[0].begin() + delta);
+    local_input_2_ = std::vector<char>(input_[1].begin(), input_[1].begin() + delta);
+  } else {
+    world.recv(0, 0, local_input_1_.data(), delta);
+    world.recv(0, 1, local_input_2_.data(), delta);
+  }
+  // Init value for output
+  res_ = 0;
```
Comment (aobolensk) [inline tasks/mpi/guseynov_e_check_lex_order_of_two_string/src/ops_mpi.cpp:88]: Please, move data transfer to ::run()

File: tasks/mpi/guseynov_e_check_lex_order_of_two_string/func_tests/main.cpp [hunk 6]
```diff
@@ -0,0 +1,397 @@
+#include <gtest/gtest.h>
+
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <vector>
+
+#include "mpi/guseynov_e_check_lex_order_of_two_string/include/ops_mpi.hpp"
+
+TEST(guseynov_e_check_lex_order_of_two_string_mpi, Test_empty_strings) {
```
Comment (aobolensk) [inline tasks/mpi/guseynov_e_check_lex_order_of_two_string/func_tests/main.cpp:9]: Please, add random strings test

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/guseynov_e_check_lex_order_of_two_string/func_tests/main.cpp (hunk 6 above)
Comment (emilien-gus) [reply tasks/mpi/guseynov_e_check_lex_order_of_two_string/func_tests/main.cpp:9]: fixed

**Review by aobolensk — APPROVED** [review]

**Review by allnes — APPROVED** [review]
